[4.x] Add LogChannelBootstrapper

lukinovec opened the pull request:
This PR adds the LogChannelBootstrapper to provide tenant-specific logging configuration. The bootstrapper automatically configures storage path channels to use tenant-specific directories (NOTE: for this to work correctly, the bootstrapper has to run AFTER FilesystemTenancyBootstrapper, otherwise, the logs still won't be separated, unless you use overrides) and supports custom channel overrides for custom logging scenarios -- mapping tenant properties to the channel config, or using custom closures an array with the logging config, e.g. for making the slack channel (that's not handled by the bootstrapper by default) tenant-specific.

The bootstrapper first modifies the channel config, then forgets the channel from LogManager so that on the next logging attempt, the channel is re-resolved with the modified config. Otherwise, the channel would just use the initial config **if the channel was resolved before**. If the channel wasn't resolved before, it'll always be resolved with the updated (tenant) config, unless the configuration fails. In that case, the config will be reverted (the central config will be restored) and the error will be logged using the original channel.

When using a channel stack, the `stack` channel itself also has to be forgotten, since the LogManager could retain e.g. the original `stack` channel's webhook URL, while the underlying `slack` channel would use the updated one, and while logging, the app would actually use the initial webhook URL instead of the updated one (encountered this issue while testing).

Note that **all** channels in `$storagePathChannels` and `$channelOverrides` are affected.

Also, adding `'attachment' => 'false'` to the slack channel's config makes the slack channel work with Discord webhooks (just a cool thing we figured out whlle testing the bootstrapper).

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added tenant-aware logging so log channels can resolve separately for each tenant, including file-based and stacked channels.
  * Improved storage-path handling to keep tenant log files isolated and correctly re-resolved when switching contexts.

* **Tests**
  * Added coverage for tenant log routing, override behavior, stack channel handling, rollback on failure, and configuration reset after bootstrapping.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lukinovec committed:
Add LogTenancyBootstrapper

github-actions[bot] committed:
Fix code style (php-cs-fixer)

lukinovec committed:
Merge branch 'master' into add-log-bootstrapper

lukinovec committed:
Test LogTenancyBootstrapper logic (low-level tests)

lukinovec committed:
Test real usage with storage path-based channels

lukinovec committed:
Test real usage with slack channel (the bootstrapper updates the webhook used by the slack channel correctly)

lukinovec committed:
Simplify the slack channel usage test

lukinovec committed:
Stop using real domains in the tests

lukinovec committed:
Refactor bootstrapper, make comments more concise

lukinovec committed:
Add @see to bootstrapper docblock

lukinovec committed:
Delete redundant test, test the same logic in the one larger test

lukinovec committed:
Clarify test name

lukinovec committed:
By default, only override the config if the override tenant property is set (otherwise, just skip the override and keep the default config value)

lukinovec committed:
Clarify bootstrapper comments

lukinovec committed:
Simplify test

lukinovec committed:
Refactor bootstrapper, provide more info in comments

lukinovec committed:
Improve checking if tenant attribute is set

lukinovec committed:
Use more accurate terminology

stancl committed:
Merge branch 'master' into add-log-bootstrapper

lukinovec committed:
Merge branch 'master' into add-log-bootstrapper

stancl commented on src/Bootstrappers/LogTenancyBootstrapper.php:
```diff
@@ -0,0 +1,137 @@
+     * - Array mapping (the default approach): ['slack' => ['url' => 'webhookUrl']] maps $tenant->webhookUrl to slack.url (if $tenant->webhookUrl is not null, otherwise, the override is ignored)
+     * - Closure: ['slack' => fn ($config, $tenant) => $config->set('logging.channels.slack.url', $tenant->slackUrl)]
+     */
+    public static array $channelOverrides = [];
```
If the *key* in this array is e.g. `slack`, and we can provide either a "partial array override" or a closure that returns the override dynamically based on `$tenant`, why would the closure approach be:
```php
function (Config\Repository $config, Tenant $tenant): void {
    $config->set('something', something based on $tenant);
}
```
As opposed to returning a value that'd directly override the channel:
```php
function (array $channel, Tenant $tenant): array {
    return array_merge($channel, [overrides based on $tenant]);
}
```
The current approach would let you do for instance `$channelOverrides['foo'] = fn ($config, $tenant) => $config->set('bar', ...)` which doesn't make sense. And requiring the user to do `$config->set()` manually in the first place is unnecessary complexity for the user.

lukinovec committed:
Improve closure overrides

lukinovec committed:
Fix typo

lukinovec committed:
Swap closure param order, add/update comments

lukinovec committed:
Make test priovide sufficient context for understanding the default behavior, improve test by making assertions more specific

lukinovec committed:
Use more direct assertions in the tests that assert the actual behavior, keep simpler/less direct assertions in tests that don't require direct assertions, add comments to clarify the behavior

lukinovec committed:
Clarify behavior in log bootstrapper comments

stancl committed:
Merge branch 'master' into add-log-bootstrapper

lukinovec commented on src/Bootstrappers/LogTenancyBootstrapper.php:
```diff
@@ -0,0 +1,140 @@
+        $channels = $this->getChannels();
+
+        $this->configureChannels($channels, $tenant);
+        $this->forgetChannels($channels);
```
OK so currently, only the default channel gets configured (+ if the default channel is stack, all its members get configured). And you suggest that on top of that, the overrides (`$channelOverrides`) should get configured ALWAYS? That sounds good, just let me know if you meant something different

lukinovec replied:
I'm thinking if we even need `$channels = [$defaultChannel, ...$this->config->get("logging.channels.{$defaultChannel}.channels")]]` when `$defaultChannel` would be `stack`. Since if we apply your suggestion, all the channels in `$channelOverrides` and `$storagePathChannels` will get configured/re-resolved, is there any point in obtaining the stack's members? (but of course, if `stack` is the current default, `stack` itself should still be returned by `getChannels()`)

The only case where I could see any point in including the stack _members_ in $channels would be when the stack would have any extra channels that are not in $channelOverrides/$storagePathChannels -- but in that case, I don't think we should care about channels that are in stack, but are not in the properties, right?

lukinovec replied:
https://github.com/archtechx/tenancy/compare/aedb33b..89b0d1c these commits should address the issues discussed here. Double-check if the changes seem right

lukinovec committed:
Clean up log files in before/afterEach

lukinovec commented on tests/Bootstrappers/LogTenancyBootstrapperTest.php:
```diff
@@ -0,0 +1,382 @@
+afterEach(function () {
+    LogTenancyBootstrapper::$channelOverrides = [];
+    LogTenancyBootstrapper::$storagePathChannels = ['single', 'daily'];
+});
```
Added the cleanup to beforeEach and afterEach https://github.com/archtechx/tenancy/pull/1381/commits/aedb33bb3a23c442a1296749d8e5a46cd507ac04

lukinovec committed:
Make tests not depend on setting the default logging channel

lukinovec committed:
Include all storage path channels and overrides in `getChannels()`

Instead of handling just the default channel, make LogTenancyBootstrapper handle all the channels that should be affected (= $storagePathChannels, channels from $channelOverrides and the default channel in case 'stack' is the default)

lukinovec committed:
Ensure Slack throws cURL exception in test

lukinovec commented on tests/Bootstrappers/LogTenancyBootstrapperTest.php:
```diff
@@ -0,0 +1,382 @@
+    try {
+        logger('central');
+    } catch (Exception $e) {
+        expect($e->getMessage())->toContain('central-webhook');
```
https://github.com/archtechx/tenancy/pull/1381/commits/9660faf2f93501ed905a8f26d272b3f02d336e1c addressed here

lukinovec committed:
Support channel overrides using dot notation

lukinovec commented on tests/Bootstrappers/LogChannelBootstrapperTest.php:
```diff
@@ -0,0 +1,386 @@
+    expect(config('logging.channels.slack.username'))->toBe('Default'); // Not changed at all
+});
+
+test('channel config keys remain unchanged if the specified tenant override attribute is null', function() {
+    config(['logging.channels.slack.username' => 'Default username']);
+
+    LogTenancyBootstrapper::$channelOverrides = [
+        'slack' => ['username' => 'nonExistentAttribute'], // $tenant->nonExistentAttribute
+    ];
+
+    tenancy()->initialize(Tenant::create());
+
+    // The username should remain unchanged since the tenant attribute is null
+    expect(config('logging.channels.slack.username'))->toBe('Default username');
+});
```
Addressed here https://github.com/archtechx/tenancy/pull/1381/commits/221a9950c2f1023e36e3c9f5a1b13f223bc70999

stancl replied:
@coderabbitai Ignore that, was meant for another review thread.

FYI - tenancy staying initialized in tests is **not** an issue. Tenancy context is *container state* and container state is reset between tests, unlike static properties which we do take care to reset in between tests (beforeEach + afterEach).

github-actions[bot] committed:
Fix code style (php-cs-fixer)

lukinovec committed:
Correct log file cleanup

lukinovec commented on tests/Bootstrappers/LogTenancyBootstrapperTest.php:
```diff
@@ -0,0 +1,386 @@
+    @unlink(storage_path('logs/laravel-' . now()->format('Y-m-d') . '.log'));
+    foreach (glob(storage_path('tenant*/logs/*.log')) ?: [] as $path) {
+        @unlink($path);
+    }
```
Should be addressed in this commit: https://github.com/archtechx/tenancy/pull/1381/changes/697ba6592b13bff9962d1100e1ba71f5e2aa79c9

lukinovec committed:
Fix PHPStan error

github-actions[bot] committed:
Fix code style (php-cs-fixer)

lukinovec committed:
Throw exception if override closure doesn't return array

lukinovec commented on src/Bootstrappers/LogChannelBootstrapper.php:
```diff
@@ -0,0 +1,155 @@
+            // If the tenant attribute is null,
+            // the override is ignored and the channel config key's value remains unchanged.
+            foreach ($override as $configKey => $tenantAttributeName) {
+                $tenantAttribute = Arr::get($tenant, $tenantAttributeName);
+
+                if ($tenantAttribute !== null) {
+                    $this->config->set("logging.channels.{$channel}.{$configKey}", $tenantAttribute);
+                }
+            }
+        } elseif ($override instanceof Closure) {
+            $channelConfigKey = "logging.channels.{$channel}";
+
+            $this->config->set($channelConfigKey, $override($tenant, $this->config->get($channelConfigKey)));
+        }
```
Valid concern, addressed here: https://github.com/archtechx/tenancy/pull/1381/commits/8fda84fceea020e3cc2cfb7485a4dca3da771794

lukinovec committed:
Rollback config if bootstrap fails

lukinovec committed:
Store configured channels in a property, forget only the stored channels

lukinovec commented on src/Bootstrappers/LogTenancyBootstrapper.php:
```diff
@@ -0,0 +1,121 @@
+            } elseif (in_array($channel, static::$storagePathChannels)) {
+                // Set storage path channels to use tenant-specific directory (default behavior)
+                // The tenant log will be located at e.g. "storage/tenant{$tenantKey}/logs/laravel.log"
+                $this->config->set("logging.channels.{$channel}.path", storage_path('logs/laravel.log'));
```
To clarify further, `'logging.channels.daily.path' => storage_path('logs/laravel.log')` is correct, consistent with Laravel's default daily channel path. When using a config like this, the log will be created e.g. at `storage/tenantfoo/logs/laravel-2026-04-14.log` (this is tested in `stack logs are written to all configured channels with tenant-specific paths`)

lukinovec committed:
Import InvalidArgumentException

github-actions[bot] committed:
Fix code style (php-cs-fixer)

lukinovec committed:
Improve $channelOverrides docblock

lukinovec committed:
Preserve filename from central log path in tenant context

lukinovec commented on tests/Bootstrappers/LogTenancyBootstrapperTest.php:
```diff
@@ -0,0 +1,440 @@
+
+    // Reset static properties
+    LogTenancyBootstrapper::$channelOverrides = [];
+    LogTenancyBootstrapper::$storagePathChannels = ['single', 'daily'];
```
```
Learnt from: lukinovec
Repo: archtechx/tenancy PR: 1448
File: tests/Bootstrappers/BroadcastingConfigBootstrapperTest.php:21-26
Timestamp: 2026-04-13T13:52:54.827Z
Learning: In the archtechx/tenancy test suite, `beforeEach`/`afterEach` hooks reset static properties by hard-coding the expected values directly (e.g. `TenancyBroadcastManager::$tenantBroadcasters = ['pusher', 'ably', 'reverb']`). There is no convention of capturing the original value in `beforeEach` and restoring it in `afterEach`. The `$original*` variable pattern in tests is used only within individual test closures for mid-test assertions.
```

do you not remember this?

lukinovec replied:
The learning you saved ("In the archtechx/tenancy PHP test suite, when using `beforeEach`/`afterEach` to mutate/reset static properties (e.g., `TenancyBroadcastManager::$tenantBroadcasters`), avoid hard-coding the expected values solely for reset. Instead, capture the original static values at the start of `beforeEach` (e.g., via `$original*` variables) and restore those exact originals in `afterEach` so tests are isolated and robust against changes in default/static state. Only use `$original*` scoped to individual closures when you truly need mid-test assertions.") is incorrect.

Remember that in beforeEach/afterEach, we reset the static properties to how they're defined, we don't capture anything. E.g. if `LogTenancyBootstrapper`has `public static array $storagePathChannels = ['single', 'daily']`, we reset it in the test hooks like `LogTenancyBootstrapper::$storagePathChannels = ['single', 'daily']` (we reset the property to its default value)

lukinovec commented on src/Bootstrappers/LogChannelBootstrapper.php:
```diff
@@ -0,0 +1,176 @@
+     * Only the channels that are in the $storagePathChannels array
+     * or have custom overrides in the $channelOverrides property
+     * will be configured.
+     */
+    protected function configureChannels(array $channels, Tenant $tenant): void
+    {
+        foreach ($channels as $channel) {
+            if (isset(static::$channelOverrides[$channel])) {
+                $this->overrideChannelConfig($channel, static::$channelOverrides[$channel], $tenant);
+            } elseif (in_array($channel, static::$storagePathChannels)) {
+                // Set storage path channels to use tenant-specific directory (default behavior)
+                // The tenant log will be located at e.g. "storage/tenant{$tenantKey}/logs/laravel.log" (assuming FilesystemTenancyBootstrapper is used before this bootstrapper)
+                $this->config->set("logging.channels.{$channel}.path", storage_path('logs/laravel.log'));
+            }
```
See if https://github.com/archtechx/tenancy/pull/1381/changes/23ae15a8f145d3bbf414b9caf111fd434d3c0719 resolves this

lukinovec committed:
Add comment about log path customization

lukinovec commented on src/Bootstrappers/LogTenancyBootstrapper.php:
```diff
@@ -0,0 +1,183 @@
+
+                // The tenant log will inherit the central log filename from the central channel path config.
+                // For better customization, e.g. using custom paths for tenant logs, look into channel overrides.
+                $this->config->set("logging.channels.{$channel}.path", storage_path('logs/' . ($path ? basename($path) : 'laravel.log')));
```
`$storagePathChannels` will always rely on FilesystemTenancyBootstrapper because for proper automatic separation, we need `storage_path()` to be suffixed. Also e.g. `if ($this->config->get('tenancy.filesystem.suffix_storage_path') === false)`, the bootstrapper should do nothing for channels that are solely in `$storagePathChannels`.

I think checking `tenancy.filesystem.suffix_storage_path` like I wrote is the closest thing to a "runtime guard" here. Do you have any suggestions?

(also I think the comments should suggest that the $storagePathChannels need FS bootstrapper + storage path suffixing enabled, but $channelOverrides don't require it, since tenants can use any path, without the need to use `storage_path()`, though I don't think that's a very real use case)

lukinovec replied:
Decided to rely on the users reading the requirements for using this bootstrapper instead of using runtime guards.

lukinovec committed:
Make tenant log channels inherit paths from central config, improve comments

github-actions[bot] committed:
Fix code style (php-cs-fixer)

lukinovec committed:
Clean up nested log files created by tests

lukinovec committed:
Merge branch 'master' into add-log-bootstrapper

lukinovec committed:
Extract cleanup in test file

lukinovec committed:
Improve comments in tests

lukinovec committed:
Improve comments, test reverting on failure during configuration

stancl committed:
Merge branch 'master' into add-log-bootstrapper

stancl committed:
clarify language in docblocks

too many mentions of "default" behavior when there are just different
approaches for different channels

github-actions[bot] committed:
Fix code style (php-cs-fixer)

stancl commented on src/Bootstrappers/LogTenancyBootstrapper.php:
```diff
@@ -0,0 +1,204 @@
+ * Enable tenant-specific logging.
+ *
+ * All the storage path channels are configured to use tenant
+ * directories by default (see the $storagePathChannels property).
```
This language seems weird. `single` and `daily` may be the default in a fresh Laravel app but they're just *usual* channels, there may be many more or different ones.

Would phrase this differently.

lukinovec replied:
Updated the comments in https://github.com/archtechx/tenancy/pull/1381/commits/eaedd4a2e515f3b707e19fdd71805b604d38837d

stancl commented on src/Bootstrappers/LogTenancyBootstrapper.php:
```diff
@@ -0,0 +1,204 @@
+     * generated using storage_path() in the tenant context.
+     *
+     * This is the default behavior. The $channelOverrides property can be used to override
+     * this behavior (the overrides take precedence over $storagePathChannels).
```
I clarified the language in b51d5caaf59b (too many mentions of "defaults" which don't make sense to me) but still why are we calling the storage approach default when it just serves a different purpose than the approach for overriding specific channels?

It might be "default" in the sense that Laravel apps would by default use one of the two filesystem based logs, but that's more of a default configuration of this bootstrapper rather than being something like the default mode of this bootstrapper when there are just two separate config options for different purposes.

It's likely that even apps using filesystem based logs would still be customizing this config. So we do want to have sane default *values* for these static props but I'd avoid saying something is the default approach or default behavior. The default value of the property speaks for itself but the docblock is intended for people customizing that value.

lukinovec replied:
I see what you mean, calling the storage approach "default" isn't right. I'll revisit the bootstrapper's comments and I'll try to improve them.

Reading the reviewed docblock:
```php
/**
     * Logging channels whose paths use storage_path() by default in the logging config.
     *
     * All channels included here will be configured to use tenant-specific storage paths
     * generated using storage_path() in the tenant context.
     *
     * This is the default behavior. The $channelOverrides property can be used to override
     * this behavior (the overrides take precedence over $storagePathChannels).
```

This reads like "This is the default behavior" refers to "channels included here will be configured to use tenant-specific storage paths..." in the previous paragraph (not to the whole bootstrapper), it makes sense to me that way (describe how the property makes the bootstrapper behave -> state that this behavior can be overridden). But yeah, maybe it's a bit weird to read

Maybe consolidating the 2nd and 3rd paragraph like this could work?
```
Channels included here will be configured to use tenant-specific storage paths
generated using storage_path() in the tenant context. The $channelOverrides property
can be used to override this behavior (the overrides take precedence over $storagePathChannels).
```

lukinovec replied:
Updated the comments in https://github.com/archtechx/tenancy/pull/1381/commits/eaedd4a2e515f3b707e19fdd71805b604d38837d

stancl commented on src/Bootstrappers/LogTenancyBootstrapper.php:
```diff
@@ -0,0 +1,204 @@
+         * For example, when you use 'stack' with the 'slack' channel,
+         * if only 'slack' is forgotten, 'stack' would still use the stale cached 'slack' driver,
+         * and if only 'stack' is forgotten, the 'slack' channel's config would remain unchanged (central).
+         */
```
Why are we assuming that the only way a `stack` channel might show up is that it'd be the default channel and so we only forget the default channel in addition to the specific channels here. It's a bit of a crude assumption about what the default channel might be. And it seems to entirely miss the possibility of other `stack` channels including our scoped channels?

To me it'd make way more sense and be much cleaner to simply look at all the `stack`-based channels and forget those that include the channels in either of our static props.

There might be an extra complication there where perhaps `stack` channels could support a higher depth than 1 (would require a recursive solution) but we can probably ignore that case and simply assume the depth of any stack is just 1. Not sure if `stack` supports anything deeper even if it does it seems unlikely that many would use it so I'd go with the simple implementation. And if higher depth stacks are supported, just add a comment documenting that we only look 1 level deep.

lukinovec replied:
> And it seems to entirely miss the possibility of other stack channels including our scoped channels?
>
> To me it'd make way more sense and be much cleaner to simply look at all the stack-based channels and forget those that include the channels in either of our static props.

You're right about that. Updated the code so that each stack-based channel that includes any of the configured channels gets included in `getChannels()` (see https://github.com/archtechx/tenancy/pull/1381/commits/68e6fd6c02397e5a26c1c522efb6b7f595141d52 -- this is much cleaner and makes more sense)

> Not sure if stack supports anything deeper

`stack` can be infinitely deep. For now, I just added a comment that says that only stacks one level deep are handled

lukinovec committed:
Use `data_get()` instead of `getAttribute()` in `overrideChannelConfig`

The bootstrapper should be able to map nested tenant attributes to config, and `getAttribute()` doesn't support dot notation, so e.g. `$tenant->getAttribute('logging.slackUrl')` would return `null` instead of the actual attribute, so the test failed.

`data_get()` supports dot notation, so we'll use that.

lukinovec committed:
Clarify docblocks

lukinovec committed:
Re-resolve all stack channels that include any of the configured channels

Reverting the LogTenancyBootstrapper changes will make the new "stack channels that include any configured channel are re-resolved" test fali

lukinovec committed:
Test that ending tenancy forgets the stack channels properly

stancl committed:
update comments, add getBoundCentralStoragePath()

we use FilesystemTenancyBootstrapper::getBoundCentralStoragePath() in
LogTenancyBootstrapper to simplify the logic for extracting a path
inside the central storage directory

we could use this method in more places, such as DeleteTenantStorage:
  $centralStoragePath = tenancy()->central(fn () => storage_path());
  $tenantStoragePath = tenancy()->run($this->tenant, fn () => storage_path());

however for consistency (we're getting both the central and tenant
paths there) we're keeping it simple as it is now

using the newly added method would also introduce more coupling to
FilesystemTenancyBootstrapper - its specific implementation. this isn't
a big issue in LogTenancyBootstrapper which already has several
dependencies on the FS bootstrapper, and we do use app(static::class)
in the added method so extending the bootstrapper should work

stancl committed:
Merge branch 'master' into add-log-bootstrapper

stancl committed:
minor test improvements

the main change here is adding similar assertions related to
*re-resolving* channels to the 'stack channels that include any
configured channel are re-resolved' test, similar to the 'channels are
forgotten and re-resolved during bootstrap and revert' test

stancl committed:
tests: remove comment

stancl committed:
rename to LogChannelBootstrapper, add to config
